crates/hotshot/src/traits/networking/push_cdn_network/definition/message_hook.rs

```diff
@@ -0,0 +1,482 @@
+        // not exceed the `global average * allowed_multiple`
+        if sample.last_checked_time.elapsed() >= self.sample_check_interval {
+            // Get our local and global bps
+            let local_bps = sample.get();
+            let mut global_bps = sma.get();
+
+            // Clamp the global bps to a minimum of 2000 if it's not zero (meaning uninitialized)
+            if global_bps > 0 {
+                global_bps = std::cmp::max(global_bps, 2000);
+            }
+
+            // Calculate the maximum allowed bps
+            let max_allowed_bps = global_bps * self.allowed_multiple;
+
+            // If the local bps is greater than the allowed bps, kick the user
+            if global_bps != 0 && local_bps > max_allowed_bps {
+                error!(
+                    "Local bps ({}) is greater than maximum allowed bps ({}), kicking user",
+                    local_bps, max_allowed_bps
+                );
+                return Err(anyhow::anyhow!(
+                    "Local bps ({}) is greater than maximum allowed bps ({}), kicking user",
+                    local_bps,
+                    max_allowed_bps
+                ));
+            }
+
+            // Reset the check time
+            sample.last_checked_time = Instant::now();
+        }
+
+        // Commit the sample if that interval has elapsed
+        if sample.last_committed_time.elapsed() >= self.sample_commit_interval {
+            sma.commit_sample(sample);
+        }
+
+        Ok(())
+    }
+
+    /// Process incoming broadcast messages from the user
+    fn process_broadcast_message(&mut self, broadcast: &mut Broadcast) -> Result<HookResult> {
+        // Process through the `SMA`
+        self.process_against_sma(broadcast.message.len(), MessageType::Broadcast)?;
+
+        // Calculate the hash of the message
+        let mut hasher = Hash64::default();
+        hasher.write(&broadcast.message);
+        hasher.write(&broadcast.topics);
+
+        // Make sure we have not already seen it
+        if self.message_hash_cache.put(hasher.finish(), ()).is_some() {
+            return Ok(HookResult::SkipMessage);
+        }
+
+        // Make sure it is deserializable
+        // let (_, _) = Self::deserialize_message(&broadcast.message)?;
+
+        Ok(HookResult::ProcessMessage)
+    }
+
+    /// Process incoming direct messages from the user
+    fn process_direct_message(&mut self, direct: &mut Direct) -> Result<HookResult> {
+        // Process through the `SMA`
+        self.process_against_sma(direct.message.len(), MessageType::Direct)?;
+
+        // Calculate the hash of the message
+        let mut hasher = Hash64::default();
+        hasher.write(&direct.message);
+        hasher.write(&direct.recipient);
+
+        // Make sure we have not already seen it
+        if self.message_hash_cache.put(hasher.finish(), ()).is_some() {
+            return Ok(HookResult::SkipMessage);
+        }
```

Comment:
nit: we are doing same hashing logic in `process_broadcast_message` and `process_direct_message` can we extract this to another method?

Reply:
not sure if itll be easy since we are doing hash on Broadcast and Direct structs, wonder if we can make them params somehow if not no worries

Reply:
Good idea, let me see if I can abstract this out